File: experiments-py/week01/day02-tool-calling/agent.py
import json
import sys
import os
from openai import OpenAI

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "../.."))
from config import api_key, MODEL, cfg


class MinAgent:
    def __init__(self, tools, functions, max_steps):
        self.client = OpenAI(base_url=cfg["base_url"], api_key=api_key)
        self.tools = tools
        self.functions = functions
        self.messages = []
        self.max_steps = max_steps

    def observe(self, user_input: str):
        self.messages.append({"role": "user", "content": user_input})

    def think(self):
        response = self.client.chat.completions.create(
            model=MODEL,
            messages=self.messages,
            tools=self.tools,
        )
        return response

    def act(self, tool_call):
        func_name = tool_call.function.name
        func_args = json.loads(tool_call.function.arguments)
        func = self.functions[func_name]
        result = func(**func_args)
        return {
            "role": "tool",
            "tool_call_id": tool_call.id,
            "content": str(result),
        }

    def run(self, user_query: str):
        self.observe(user_query)

        for _ in range(self.max_steps):
            response = self.think()
            message = response.choices[0].message
            self.messages.append(message)

            if not message.tool_calls:
                return message.content

            for tool_call in message.tool_calls:
                tool_result = self.act(tool_call)
                self.messages.append(tool_result)

        return "达到最大执行轮次，任务停止"

File: experiments-.net/Day02.ToolCalling/Agent.cs
using AiAgent.Shared;

namespace Day02.ToolCalling;

public class Agent
{
    private readonly ILlmClient _llm;
    private readonly List<ChatMessage> _messages = new();
    private readonly int _maxSteps;

    public Agent(ILlmClient llm, int maxSteps = 10)
    {
        _llm = llm;
        _maxSteps = maxSteps;
    }

    public void Observe(string userInput)
    {
        _messages.Add(new ChatMessage { Role = "user", Content = userInput });
    }

    public async Task<string> RunAsync(string userQuery)
    {
        Observe(userQuery);

        for (var i = 0; i < _maxSteps; i++)
        {
            var response = await _llm.ChatWithToolsAsync(_messages, Tools.Definitions);

            _messages.Add(new ChatMessage
            {
                Role = "assistant",
                Content = response.Content ?? "",
                ToolCalls = response.ToolCalls
            });

            if (response.ToolCalls == null || response.ToolCalls.Count == 0)
                return response.Content ?? "";

            foreach (var toolCall in response.ToolCalls)
            {
                Console.WriteLine($"  ▶ 调用工具: {toolCall.FunctionName}({toolCall.FunctionArguments})");

                if (Tools.Functions.TryGetValue(toolCall.FunctionName, out var func))
                {
                    var result = func(toolCall.FunctionArguments);
                    _messages.Add(new ChatMessage
                    {
                        Role = "tool",
                        ToolCallId = toolCall.Id,
                        Content = result
                    });
                }
            }
        }

        return "达到最大执行轮次，任务停止";
    }
}
